File: torchsl/csrc/ops/subspace_learning/pca.cpp
#include <vector>
#include <tuple>
#include <torch/script.h>
#include "singleview_helpers.h"

namespace torchsl {
std::tuple<torch::Tensor, torch::Tensor> pca(
    const torch::Tensor& X
    ) {
    const auto options = X.options().requires_grad(false);
    const int num_samples = X.size(0);
    const int dim = X.size(1);

    auto X_centered = torchsl::centerize(X);
    auto cov = X_centered.t().mm(X_centered).div_(num_samples - 1);
    auto SI = torch::eye(dim, options);

    return std::make_tuple(SI, cov);
}
}

File: torchsl/ops/subspace_learning/pca.py
import torch

import torchsl
from torchsl._extensions import _has_ops

__all__ = ['pca']


# ===========================================================
# Principle Component Analysis
# ===========================================================
def pca(X):
    if _has_ops():
        return torchsl.ops.pca(X)

    options = dict(dtype=X.dtype, device=X.device)
    num_samples = X.size(0)
    dim = X.size(1)
    X_centered = X - X.mean(0)
    cov = X_centered.t().mm(X_centered).div(num_samples - 1)
    SI = torch.eye(dim, **options)

    return SI, cov
